File: Python/boxes.py
def main():
    n = int(input())
    tree = [[] for _ in range(n+1)]
    boxes = list(map(int, input().split(' ')))

    for i in range(len(boxes)):
        if boxes[i]:
            tree[boxes[i]].append(i+1)

    for _ in range(int(input())):
        queries = list(map(int, input().split(' ')))[1:]
        visited = set()

        while queries:
            curr = queries.pop()
            visited.add(curr)
            for b in tree[curr]:
                if b not in visited:
                    queries.append(b)

        print(len(visited))


if __name__ == '__main__':
    main()

File: Kotlin/boxes.kt
fun main() {
    val n = readLine()!!.toInt()
    val tree = mutableListOf<MutableList<Int>>()
    val boxes = readLine()!!.split(' ').map(String::toInt)

    for (i in 0..n) {
        tree.add(mutableListOf<Int>())
    }

    for (i in 0..boxes.size-1) {
        if (boxes[i] > 0) {
            tree[boxes[i]].add(i+1)
        }
    }

    for (i in 1..readLine()!!.toInt()) {
        val tmp = readLine()!!.split(' ').map(String::toInt)
        val queries = tmp.subList(1, tmp.size).toMutableList()
        val visited = mutableSetOf<Int>()

        while (queries.isNotEmpty()) {
            val curr = queries.removeLast()
            visited.add(curr)

            for (b in tree[curr]) {
                if (b !in visited) {
                    queries.add(b)
                }
            }
        }
        println(visited.size)
    }
}